File: src/StormyCommerce.Core/Models/Dtos/GatewayResponses/Catalog/ProductDto.cs
﻿using StormyCommerce.Core.Entities.Catalog.Product;
using StormyCommerce.Core.Entities.Media;
using StormyCommerce.Module.Catalog.Dtos;
using System;
using System.Collections.Generic;
using System.Linq;

namespace StormyCommerce.Core.Models.Dtos.GatewayResponses.Catalog
{
    public class ProductDto
    {
        public ProductDto(){}

        public ProductDto(StormyProduct product, long id)
        {
            Id = id;
            ProductName = product.ProductName;
            Slug = product.Slug;
            QuantityPerUnity = product.QuantityPerUnity;
            AvailableSizes = AvailableSizes;
            Height = product.Height;
            Width = product.Width;
            Length = product.Length;
            Discount = product.Discount;
            UnitWeight = product.UnitWeight;
            UnitsInStock = product.UnitsInStock;
            UnitsOnOrder = product.UnitsOnOrder;
            Price = product.Price;                  
        }

        public ProductDto(StormyProduct product)
        {
            Id = product.Id;
            ProductName = product.ProductName;
            Slug = product.Slug;
            QuantityPerUnity = product.QuantityPerUnity;
            AvailableSizes = product.AvailableSizes;
            Height = product.Height;
            Width = product.Width;
            Length = product.Length;
            Discount = product.Discount;
            UnitWeight = product.UnitWeight;
            UnitsInStock = product.UnitsInStock;
            UnitsOnOrder = product.UnitsOnOrder;
            Price = product.Price;            
            ThumbnailImage = product.ThumbnailImage;
            Medias = product.Medias.Select(m => new ProductMediaDto(m)).ToList();
            Categories = product.Categories.Select(c => new ProductCategoryDto(c)).ToList();
            Brand = new BrandDto(product.Brand);                        

            Vendor = new VendorDto(product.Vendor);
        }

        public long Id { get; set; }
        public string ProductName { get; set; }
        public string ShortDescription { get; set; }
        public bool ProductAvailable { get; set; }
        public string Slug { get; set; }
        public int QuantityPerUnity { get; set; }
        public string AvailableSizes { get; set; }
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; }
        public double Height { get; set; }
        public double Width { get; set; }
        public double Length { get; set; }
        public double UnitWeight { get; set; }
        public int UnitsInStock { get; set; }
        public int UnitsOnOrder { get; set; }
        public string Price { get; set; }
        public string OldPrice { get; set; }
        public string ThumbnailImage { get; set; }
        public List<ProductCategoryDto> Categories { get; set; } = new List<ProductCategoryDto>();
        public BrandDto Brand { get; set; }
        public VendorDto Vendor { get; set; }
        public List<ProductMediaDto> Medias { get; set; }

        public StormyProduct ToStormyProduct()
        {
            return new StormyProduct
            {
                ProductName = this.ProductName,
                Slug = this.Slug,
                QuantityPerUnity = this.QuantityPerUnity,
                AvailableSizes = this.AvailableSizes,
                UnitPrice = this.UnitPrice,
                Discount = this.Discount,
                UnitWeight = this.UnitWeight,
                UnitsInStock = this.UnitsInStock,
                UnitsOnOrder = this.UnitsOnOrder,
                Price = this.Price,                        
            };
        }
    }
}

File: Web/models/catalog/ProductDto.ts
import { CategoryDto } from './categoryDto';
import { VendorDto } from './vendorDto';
import { BrandDto } from './vendorDto';
import { MediaDto } from '../common/mediaDto';

export class ProductDto implements IProductDto {
    id?: number | undefined;
    productName?: string | undefined;
    slug?: string | undefined;
    quantityPerUnity?: number | undefined;
    unitSize?: string | undefined;
    unitPrice?: number | undefined;
    discount?: number | undefined;
    unitWeight?: number | undefined;
    unitsInStock?: number | undefined;
    unitsOnOrder?: number | undefined;
    price?: string | undefined;
    oldPrice?: string | undefined;
    thumbnailImage?: string | undefined;
    category?: CategoryDto | undefined;
    brand?: BrandDto | undefined;
    vendor?: VendorDto | undefined;
    medias?: MediaDto[] | undefined;

    constructor(data?: IProductDto) {
        if (data) {
            for (var property in data) {
                if (data.hasOwnProperty(property))
                    (<any>this)[property] = (<any>data)[property];
            }
        }
    }

    init(data?: any) {
        if (data) {
            this.id = data["id"];
            this.productName = data["productName"];
            this.slug = data["slug"];
            this.quantityPerUnity = data["quantityPerUnity"];
            this.unitSize = data["unitSize"];
            this.unitPrice = data["unitPrice"];
            this.discount = data["discount"];
            this.unitWeight = data["unitWeight"];
            this.unitsInStock = data["unitsInStock"];
            this.unitsOnOrder = data["unitsOnOrder"];
            this.price = data["price"];
            this.oldPrice = data["oldPrice"];
            this.thumbnailImage = data["thumbnailImage"];
            this.category = data["category"] ? CategoryDto.fromJS(data["category"]) : <any>undefined;
            this.brand = data["brand"] ? BrandDto.fromJS(data["brand"]) : <any>undefined;
            this.vendor = data["vendor"] ? VendorDto.fromJS(data["vendor"]) : <any>undefined;
            if (Array.isArray(data["medias"])) {
                this.medias = [] as any;
                for (let item of data["medias"])
                    this.medias!.push(MediaDto.fromJS(item));
            }
        }
    }

    static fromJS(data: any): ProductDto {
        data = typeof data === 'object' ? data : {};
        let result = new ProductDto();
        result.init(data);
        return result;
    }

    toJSON(data?: any) {
        data = typeof data === 'object' ? data : {};
        data["id"] = this.id;
        data["productName"] = this.productName;
        data["slug"] = this.slug;
        data["quantityPerUnity"] = this.quantityPerUnity;
        data["unitSize"] = this.unitSize;
        data["unitPrice"] = this.unitPrice;
        data["discount"] = this.discount;
        data["unitWeight"] = this.unitWeight;
        data["unitsInStock"] = this.unitsInStock;
        data["unitsOnOrder"] = this.unitsOnOrder;
        data["price"] = this.price;
        data["oldPrice"] = this.oldPrice;
        data["thumbnailImage"] = this.thumbnailImage;
        data["category"] = this.category ? this.category.toJSON() : <any>undefined;
        data["brand"] = this.brand ? this.brand.toJSON() : <any>undefined;
        data["vendor"] = this.vendor ? this.vendor.toJSON() : <any>undefined;
        if (Array.isArray(this.medias)) {
            data["medias"] = [];
            for (let item of this.medias)
                data["medias"].push(item.toJSON());
        }
        return data; 
    }
}

export interface IProductDto {
    id?: number | undefined;
    productName?: string | undefined;
    slug?: string | undefined;
    quantityPerUnity?: number | undefined;
    unitSize?: string | undefined;
    unitPrice?: number | undefined;
    discount?: number | undefined;
    unitWeight?: number | undefined;
    unitsInStock?: number | undefined;
    unitsOnOrder?: number | undefined;
    price?: string | undefined;
    oldPrice?: string | undefined;
    thumbnailImage?: string | undefined;
    category?: CategoryDto | undefined;
    brand?: BrandDto | undefined;
    vendor?: VendorDto | undefined;
    medias?: MediaDto[] | undefined;
}